lib/mini_sql/postgres/deserializer_cache.rb

```diff
@@ -4,6 +4,87 @@ module MiniSql
   module Postgres
     class DeserializerCache
 
+      class SerializableMaterializer < Array
+        attr_accessor :materializer, :decorator_module
+
+        def initialize(materializer:, decorator_module: nil)
+          @materializer = materializer
+          @decorator_module = decorator_module
+        end
+
+        def self.build_materializer(fields:)
+          i = 0
+          while i < fields.length
+            # special handling for unamed column
+            if fields[i] == "?column?"
+              fields[i] = "column#{i}"
+            end
+            i += 1
+          end
+
+          Class.new do
+            attr_accessor(*fields)
+
+            # AM serializer support
+            alias :read_attribute_for_serialization :send
+
+            def to_h
+              r = {}
+              instance_variables.each do |f|
+                r[f.to_s.sub('@', '').to_sym] = instance_variable_get(f)
+              end
+              r
+            end
+
+            def self.materialize_from_hash(result_hash)
+              r = self.new
+              result_hash.each do |field, value|
+                r.send(:"#{field}=", value)
+              end
+              r
+            end
+
+            instance_eval <<~RUBY
+              def materialize(pg_result, index)
+                r = self.new
+                #{col = -1; fields.map { |f| "r.#{f} = pg_result.getvalue(index, #{col += 1})" }.join("; ")}
+                r
+              end
+            RUBY
+          end
+        end
+
+        def materialize(result)
+          i = 0
+          # quicker loop
+          while i < result.ntuples
+            self << materializer.materialize(result, i)
+            i += 1
+          end
+          self
+        end
+
+        def marshal_dump
+          {
+            result: map(&:to_h),
```

Comment:
I think you can have the fields and then just do a to_a on each row so you don't store column names per row.

Reply:
ok, done